src/main/java/org/prebid/server/bidder/grid/GridKeywordsUtil.java

```diff
@@ -0,0 +1,244 @@
+                ? KeywordSegment.of(name, value)
+                : null;
+    }
+
+    public static List<KeywordSegment> resolveAlternativePublisherSegments(JsonNode publisherValueNode,
+                                                                           JacksonMapper mapper) {
+
+        return jsonNodeToMap(publisherValueNode, mapper).entrySet().stream()
+                .sorted(Map.Entry.comparingByKey())
+                .filter(GridKeywordsUtil::isValidPublisherEntry)
+                .flatMap(GridKeywordsUtil::mapPublisherEntryToKeywordsStream)
+                .collect(Collectors.toList());
+    }
+
+    private static boolean isValidPublisherEntry(Map.Entry<String, JsonNode> publisherEntry) {
+        final JsonNode publisherEntryValue = publisherEntry.getValue();
+        return publisherEntryValue != null && publisherEntryValue.isArray();
+    }
+
+    private static Stream<KeywordSegment> mapPublisherEntryToKeywordsStream(
+            Map.Entry<String, JsonNode> publisherEntry) {
+
+        final List<KeywordSegment> keywordSegments = new ArrayList<>();
+        final Iterator<JsonNode> publisherEntryElements = publisherEntry.getValue().elements();
+
+        while (publisherEntryElements.hasNext()) {
+            final JsonNode currentNode = publisherEntryElements.next();
+            if (currentNode.isTextual()) {
+                keywordSegments.add(KeywordSegment.of(publisherEntry.getKey(), currentNode.asText()));
+            }
+        }
+        return keywordSegments.stream();
+    }
+
+    public static Keywords merge(JacksonMapper mapper, Keywords... keywords) {
+        return Keywords.of(
+                mergeSections(extractSections(Keywords::getUser, keywords), mapper),
+                mergeSections(extractSections(Keywords::getSite, keywords), mapper));
+    }
+
+    public static Stream<ObjectNode> extractSections(Function<Keywords, ObjectNode> extractor, Keywords... keywords) {
+        return Arrays.stream(keywords)
+                .map(keyword -> clearObjectNode(ObjectUtil.getIfNotNull(keyword, extractor)))
+                .filter(Objects::nonNull);
+    }
+
+    private static ObjectNode mergeSections(Stream<ObjectNode> sections, JacksonMapper mapper) {
+        return sections.reduce(
+                mapper.mapper().createObjectNode(),
+                (left, right) -> (ObjectNode) mergeSections(left, right));
+    }
+
+    public static JsonNode mergeSections(JsonNode mainNode, JsonNode updateNode) {
+        final Iterator<String> updateFieldNames = updateNode.fieldNames();
+        while (updateFieldNames.hasNext()) {
+            final String updateFieldName = updateFieldNames.next();
+            final JsonNode valueToBeUpdated = mainNode.get(updateFieldName);
+            final JsonNode updateValue = updateNode.get(updateFieldName);
+
+            if (valueToBeUpdated != null && valueToBeUpdated.isArray() && updateValue.isArray()) {
```

Comment:
Can updateValue be null? Also create method for `valueToBeUpdated != null && valueToBeUpdated.isArray()` there are similar checks to this